pkg/cache/metrics.go

```diff
@@ -0,0 +1,107 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// BaseMetricsCollector defines the interface for recording cache metrics.
+// Each method corresponds to a specific cache-related operation.
+type BaseMetricsCollector interface {
+	RecordHit(cacheName string)
+	RecordMiss(cacheName string)
+	RecordSet(cacheName string)
+	RecordDelete(cacheName string)
+	RecordClear(cacheName string)
+}
+
+// MetricsCollector encapsulates all Prometheus metrics with labels.
+// It holds Prometheus counters for cache operations, which help track
+// the performance and usage of the cache.
+type MetricsCollector struct {
+	// Base metrics.
+	hits    *prometheus.CounterVec
+	misses  *prometheus.CounterVec
+	sets    *prometheus.CounterVec
+	deletes *prometheus.CounterVec
+	clears  *prometheus.CounterVec
+}
+
+var (
+	collectorOnce sync.Once // Ensures that the collector is initialized only once.
+	collector     *MetricsCollector
+)
+
+// InitializeMetrics initializes the singleton MetricsCollector.
+// It sets up Prometheus counters for cache operations (hits, misses, sets, deletes, clears).
+// Should be called once at the start of the application.
+func InitializeMetrics(namespace, subsystem string) {
+	collectorOnce.Do(func() {
+		collector = &MetricsCollector{
```

Comment:
Consider having this called automatically when the package is imported using init()?

Reply:
Yea, that makes sense. Good call.

Reply:
Also cleans up some of the test code.

Reply:
Another option is to just declar them as package vars like we do with most others